bhm.8
.TH "bhm" "8" "0.70" "russell@coker.com.au" "Postal"
.SH "NAME"
bhm \- program to send SMTP mail to /dev/null

.SH "SYNOPSIS"
.B bhm
.I [\-m maximum\-message\-size] [\-t threads] [\-p listen\-port] [ \-s ]
.I [\-a] [\-[z|Z] debug\-file] [\-r reverse\-dns] user\-list\-filename

.SH "DESCRIPTION"
This manual page documents briefly the
.BR bhm ,
program.
.P
It is designed to test the performance of SMTP relay servers by receiving mail
to /dev/null at random speeds.
.P
The
.B user\-list\-filename
is the name of a file which contains a list of valid user's email addresses.
The file may have three fields separated by spaces, the email address, the
password (used for rabid) and an optional third field to specify what is to be
done with the mail.  A 'd' character means that it is to be deferred
(SMTP 45x), a 'r' means that it is to be rejected (SMTP 55x), a 'b' means that
it is to be accepted and bounced, and a 'g' means that it is to be grey-listed.
.P
The
.B reverse\-dns
file is in the same format as /etc/hosts and gives the value that will be
used in the 220 message.  Otherwise it will be
.B bhm
.P
The
.B maximum\-message\-size
indicates the size in Kilobytes that will be the maximum size of a message
that will be accepted.  The default is 10240k.
.P
The
.B \-s
parameter enables SSL.
.P
The
.B listen\-port
parameter specifies the TCP port to listen on (default 25).
.P
The
.B threads
parameter is the number of threads that may be created to receive mail.
Note that your BHM machine may have hardware and OS limits that determine
how many connections it may receive.
.P
The
.B \-a
command turns on all logging.  All message data received will be logged.  This
will make it slow and it may not be able to saturate a fast Ethernet link...
.P
The
.B \-z
switch allows you to specify a debugging file base.  From this base one file
is created for each thread (with a ':' and the thread number appended), each
file is used to log all IO performed by that thread for debugging purposes.
.P
The
.B \-Z
switch is the same but creates a separate file for each connection as well
with an attitional ':' appended followed by the connection number.

.SH "RETURN CODES"
.TP
.B 0
No Error
.TP
.B 1
Bad Parameters
.TP
.B 2
System Error, lack of memory or some other resource

.SH "AUTHOR"
This program, it's manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.

.SH "AVAILABILITY"
The source is available from http://doc.coker.com.au/projects/postal/ .
.P
See http://etbe.coker.com.au/category/benchmark for further information.

.SH "SEE ALSO"
.BR postal (8), postal\-list (8), rabid (8)


debian/changelog
postal (0.72) unstable; urgency=low

  * Make LMTP work.
  * Accept a tab as a field delimiter for the user-list file.

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jan 2012 19:27:21 +1100

postal (0.71) unstable; urgency=low

  * Made rabid get the password right when -u is used.  Thanks to Lars Nordin
    for the patch.
  * Made rabid correctly check for the '.' at the end of the results of the
    list command.
  * Made all the programs correctly handle a userlist that has empty lines.
  * Use the -lgcrypt option when using gcrypt.
  * Only call gcry_control() if GNUTLS_VERSION_NUMBER <= 0x020b00
  * Ignore blank lines in user list.
  * Add subject, date, and msgid to the md5sum in the client.
    Closes: #611907
  * Stop using gnutls_kx_set_priority()
    Closes: #624038
  * Added a patch from Michael Brown <michael@netdirect.ca> to make an RPM build
    on RHEL5 (and probably other RPM systems).
  * Added a patch from Brad Hards <bradh@frogmouth.net> to fix typos in the
    man page.
  * Rewrote a lot of the md5 checking code, now the md5 hash is stored in a
    header.  It should work now.

 -- Russell Coker <russell@coker.com.au>  Mon, 02 Jan 2012 17:08:50 +1100

postal (0.70) unstable; urgency=low

  * Made it build correctly when no SSL support is configured.  Also tidy
    up the documentation.

 -- Russell Coker <russell@coker.com.au>  Mon, 26 May 2008 19:33:48 +1000

postal (0.69) unstable; urgency=low

  * Really fix the GCC 4.3 issues.  The main change was to remove the
    OS/2 and Win32 support (which probably didn't work anyway).  The previous
    version actually fixed GCC 4.2 issues.  I don't plan to fix the
    ext/hash_map issue yet, it doesn't stop anything from working and I'm
    afraid of breaking things for the majority of systems which use older
    headers.  Thanks to Bradley Smith and Chris Lamb for sending a patch.
    Closes: #474831
    Thanks to Martin Michlmayr for sending a patch.
    Closes: #417480

 -- Russell Coker <russell@coker.com.au>  Sat, 19 Apr 2008 08:07:00 +1000

postal (0.68) unstable; urgency=low

  * Added $(DESTDIR) support to the Makefile and -lrt for nanosleep() on
    OpenSolaris (the OpenSolaris code is not tested).
  * Fixed a minor compilation bug when not using GNUTLS.
    Thanks Frank Fegert <frank.fegert@web.de>.
  * Made it not use C++ warning flags on C compilation.
  * Updated it to the latest Debian standards for Lenny and compile with
    GCC 4.3. Closes: #474831

 -- Russell Coker <russell@coker.com.au>  Wed,  9 Apr 2008 11:16:00 +1000

postal (0.67-0.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with gcc-4.3. Thanks to MArtin Michlmayr. Closes: #417480

 -- Andreas Barth <aba@not.so.argh.org>  Sun, 16 Mar 2008 23:48:44 +0000

postal (0.67) unstable; urgency=low

  * Changed license to GPLv3.

 -- Russell Coker <russell@coker.com.au>  Mon,  2 Jul 2007 20:12:00 +1000

postal (0.66) unstable; urgency=low

  * Added MessageId to Postal.
  * Added "make dep".
  * Made GNUTLS work in bhm.

 -- Russell Coker <russell@coker.com.au>  Fri,  2 Mar 2007 18:44:00 +1100

postal (0.65) unstable; urgency=low

  * Added new program bhm to listen on port 25 and send mail to /dev/null.
    This allows testing mail relay systems.
  * Fixed a minor bug in reporting when compiled without SSL.
  * Made postal write the date header in correct RFC2822 format.
  * Removed the name-expansion feature, it confused many people and is not
    needed now that desktop machines typically have 1G of RAM.
    Now postal and rabid can have the same user-list file.
  * Moved postal-list into the bin directory.
  * Changed the thread stack size to 32K (used to be the default of 10M) to
    save virtual memory size (not that this makes much difference to
    anything other than the maximum number of threads on i386).
  * Added a minimum message size option to Postal.
  * Added a Postal option to specify a list of sender addresses separately to
    the list of recipient addresses.
  * Removed some unnecessary error messages.
  * Handle EINTR to allow ^Z and "bg" from the command line.  I probably
    don't handle all cases, but now that I agree that failure to handle ^Z is
    an error I expect bug reports.
  * Made the test programs display output on the minute, previously they
    displayed once per minute (EG 11:10:35) while now it will be 11:10:00.
    This also means that the first minute reported will have something less
    than 60 seconds of data - this does not matter as a mail server takes
    longer than that to get up to speed.
  * Added support for GNUTLS and made the Debian package build with it.
    Note that BHM doesn't yet work correctly with TLS.
  * Made the programs exit cleanly.

 -- Russell Coker <russell@coker.com.au>  Thu, 18 Jan 2007 08:14:00 +1100

postal (0.62.1) unstable; urgency=low

  * NMU
  * Build-Depend: libssl-dev, not libssl096-dev.  Closes: #176956

 -- LaMont Jones <lamont@debian.org>  Sun, 25 May 2003 21:33:07 -0600

postal (0.62) unstable; urgency=high

  * Added an INSTALL document.

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Jan 2003 13:28:00 +0100

postal (0.61-1) unstable; urgency=high

  * Changed the section to main.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 23:17:00 +0200

postal (0.61) unstable; urgency=high

  * Fixed the makefile to not use "-d" or "-p" options to $(INSTALL).

  * Changed connect() functions to Connect() because of some problems with
    Solaris.

  * Removed the optind=0 which caused problems on BSD.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 21:16:00 +0200

postal (0.60) unstable; urgency=high

  * Added #include <stdlib.h> to mutex.cpp to make exit() be resolved correctly.

  * Fixed the delete percentage code for rabid.  The option "-d100:0" will mean
    that no messages get deleted.

  * Added a new version of the thread code from Bonnie++, fixes a few bugs.


bhm.cpp
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include "bhmusers.h"
#include <errno.h>
#include <ctype.h>
#include <unistd.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdio.h>
#include <sys/poll.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/ip.h>

#include "postal.h"
#include "logit.h"
#include "results.h"
#include "basictcp.h"
#ifdef USE_GNUTLS
#include <errno.h>
#include <gcrypt.h>
GCRY_THREAD_OPTION_PTHREAD_IMPL;
#endif

int processes = 0;
int *thread_status;

void usage(CPCCHAR msg = NULL)
{
  if(msg)
    printf("Error: %s\n\n", msg);

  printf("Usage: bhm [-m maximum-message-size] [-t threads] [-p port]"
#ifdef USE_SSL
         " [-s]"
#endif
         "\n"
         "              [-a] [-(z|Z) debug-file] user-list-filename\n"
         "\n"
         "Postal Version: " VER_STR "\n");
  close(1);
  exit(eParam);
}

int maxMsgSize = 10240;
results res;
Logit *log;

int exitCount = 0;

void endit(int)
{
  exitCount++;
  if(exitCount > 2)
    exit(1);
}

typedef struct
{
  int threadNum;
  int fd;
  struct sockaddr_in addr;
  int ssl;
  Logit *debug;
} thread_data;

enum { eFree = 0, eUsed, eFinished };

int check_sender(CPCCHAR addr)
{
  return 0;
}

int readCommand(base_tcp &t, char *buf, int bufSize, bool stripCR, int timeout = 60);

// return 1 for successfully accepting an address, 0 for quit, <0 for error
// return 2 for data
int recv_addr(base_tcp &t, char *buf, const size_t buf_len, CPCCHAR msg, string &addr, bool do_data)
{
  while(1)
  {
    char *tmp;
    bool syntax_warn = false;
    int len = readCommand(t, buf, buf_len, true);
    if(len < 0)
      return len;
    if(!strncasecmp(buf, "quit", 4) && (isblank(buf[4]) || buf[4] == '\0'))
      return 0;
    if(do_data && !strncasecmp(buf, "data", 4) && (len == 4 || (isblank(buf[4])  && len == 5)))
      return 2;
    if(!strncasecmp(buf, msg, strlen(msg)))
    {
      CPCCHAR bad_sender_msg = "501 Bad address syntax\r\n";
      if(buf[len - 1] == '\r')
      {
        len--;
        buf[len] = '\0';
      }
      tmp = buf + strlen(msg);
      while(isblank(*tmp))
        tmp++;
      if(*tmp != '<' && buf[len - 1] != '>')
      {
        syntax_warn = true;
      }
      else if(*tmp != '<' || buf[len - 1] != '>')
      {
        t.sendCstr(bad_sender_msg);
        res.error();
        continue;
      }
      else
      {
        tmp++;
        len--;
        buf[len] = '\0';
      }
      if(check_sender(tmp))
      {
        t.sendCstr(bad_sender_msg);
        res.error();
        continue;
      }
      addr = tmp;
      if(syntax_warn)
        t.sendCstr("250 Ok but you should use <addr>\r\n");
      else
        t.sendCstr("250 Ok\r\n");
      break;
    }
    res.error();
    int rc = t.sendCstr("502 Error: Command not recognised or not appropriate at this time\r\n");
    if(rc < 0)
      return rc;
  }
  return 1;
}

// return 1 for successfully accepting a message, 0 for quit, -1 for error
int recv_msg(base_tcp &t, char *buf, const size_t buf_len)
{
  string sender;
  string recipient;
  int rc = recv_addr(t, buf, buf_len, "mail from:", sender, false);
  if(rc != 1)
    return rc;

  int recipient_count = 0;
  while( (rc = recv_addr(t, buf, buf_len, "rcpt to:", recipient, recipient_count)) == 1)
  {
    recipient_count++;
  }
  if(rc <= 0)
    return rc;
  t.sendCstr("354 End data with <CR><LF>.<CR><LF>\r\n");
  while( (rc = readCommand(t, buf, buf_len, true)) >= 0)
  {
    if(!strcmp(".", buf))
      break;
  }
  if(rc < 0)
  {
    res.error();
    return rc;
  }
  struct timespec req;
  req.tv_sec = 0;
  req.tv_nsec = 1000000;
  nanosleep(&req, NULL);
  t.sendCstr("250 Ok: queued on /dev/null\r\n");
  res.message();
  return 1;
}

const char *hostname = "bhm.coker.com.au";

int do_helo(base_tcp &t, char *buf, const size_t buf_len)
{
  int len;
  while(1)
  {
    int rc;
    if( (len = t.readLine(buf, buf_len, true)) < 0)
    {
      res.error();
      return -1;
    }
    if(buf[len - 1] == '\r')
    {
      len--;
      buf[len] = '\0';
    }
    bool basicHello = false;
    if(!strncasecmp("quit", buf, 4))
    {
      t.sendCstr("221 Bye\r\n");
      return -2;
    }
    if(!strncasecmp("helo", buf, 4))
    {
      basicHello = true;
    }
    else if(strncasecmp("ehlo", buf, 4))
    {
      res.error();
      if(t.sendCstr("502 Error: command not recognized\r\n") < 0)
        return -3;
      continue;
    }
    if(len < 6 || !isblank(buf[4]))
    {
      res.error();
      if(t.sendCstr("501 Syntax: EHLO hostname or HELO hostname\r\n") < 0)
        return -3;
      continue;
    }
    string remoteName(buf + 5);
    if(basicHello)
      rc = t.printf("250 %s\r\n", hostname);
    else
      rc = t.printf("250-%s\r\n250-PIPELINING\r\n"
#ifdef USE_SSL
                    "250-STARTTLS\r\n"
#endif
                    "250 SIZE %d\r\n", hostname, maxMsgSize * 1024);
    if(rc < 0)
    {
      res.error();
      return -1;
    }
    break;
  }
  return 0;
}

int readCommand(base_tcp &t, char *buf, int bufSize, bool stripCR, int timeout)
{
#ifdef USE_SSL
  if(!t.isTLS())
  {
    int rc = t.readLine(buf, bufSize, stripCR, timeout);
    if(rc < 0 || strcasecmp(buf, "starttls"))
      return rc;
    t.printf("220 2.0.0 Ready to start TLS\r\n");
    if(t.ConnectTLS())
      return -1;
    res.connect_ssl();
    rc = do_helo(t, buf, bufSize);
    if(rc < 0)
      return rc; // need to make sure values are consistent
    rc = t.readLine(buf, bufSize, stripCR, timeout);
    return rc;
  }
  else
#endif
  return t.readLine(buf, bufSize, stripCR, timeout);
}

void do_work(thread_data *td)
{
  base_tcp t(td->fd, log, td->debug, &res
#ifdef USE_SSL
    , td->ssl
#endif
  );
  struct sockaddr_in name;
  socklen_t namelen = sizeof(name);
  char buf[1024];

  int rc = getsockname(td->fd, (sockaddr *)&name, &namelen);

  if(rc)
  {
    fprintf(stderr, "Error getting socket name.\n");
    return;
  }
/*  if(!inet_ntop(name.sin_family, &name.sin_addr, buf, sizeof(buf)))
    printf("error decoding\n");
  else
    printf("Addr:%s\n", buf);
*/
  t.printf("220 %s ESMTP BHM\r\n", hostname);
  do_helo(t, buf, sizeof(buf));
  while( (rc = recv_msg(t, buf, sizeof(buf))) == 1)
    { }
  if(rc == eTimeout)
    t.printf("421 %s Error: timeout exceeded\r\n", hostname);
  if(rc < 0)
  {
    res.error();
    return;
  }
  if(rc == 0)
    t.sendCstr("221 2.0.0 Bye\r\n");
}

PVOID smtp_worker(PVOID param)
{
  thread_data *td = (thread_data *)param;
  do_work(td);
  if(td->debug)
    delete td->debug;
  thread_status[td->threadNum] = eFinished;
  close(td->fd);
  free(td);
  return 0;
}

int main(int argc, char **argv)
{
  int maxThreads = 100;
  bool allLog = false;
  PCCHAR debugName = NULL;
  bool debugMultipleFiles = false;
#ifdef USE_SSL
  int use_ssl = false;
#endif
  unsigned short port = 25;
#ifdef USE_GNUTLS
  gcry_control(GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
  gnutls_global_init();
  if(!gnutls_check_version(GNUTLS_VER))
  {
    fprintf(stderr, "Needs version " GNUTLS_VER " of GNUTLS\n");
    exit(1);
  }
#endif

  int c;
  while(-1 != (c = getopt(argc, argv, "m:p:st:z:Z:")) )
  {
    switch(char(c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'a':
        allLog = true;
      break;
      case 'm':
        maxMsgSize = atoi(optarg);
      break;
      case 'p':
        port = atoi(optarg);
      break;
      case 's':
#ifdef USE_SSL
        use_ssl = true;
#else
        usage("SSL not supported.\n");
#endif
      break;
      case 't':
        maxThreads = atoi(optarg);
      break;
      case 'Z':
        debugMultipleFiles = true;
      case 'z':
        debugName = optarg;
      break;
    }
  }
  if(maxMsgSize < 0 || maxMsgSize > MAX_MSG_SIZE)
    usage("Bad maximum message size.\n");

  if(maxThreads < 0 || maxThreads > MAX_PROCESSES)
    usage("Bad maximum process limit.\n");

  if(optind + 1 != argc)
    usage();

  BHMUsers ul(argv[optind]);

/*  int fd[2];
  if(pipe(fd))
  {
    printf("Can't create pipe.\n");
    return eSystem;
  }*/
  struct sigaction sa;
  sa.sa_sigaction = NULL;
  sa.sa_handler = SIG_IGN;
  sa.sa_flags = 0;
  if(sigaction(SIGPIPE, &sa, NULL))
  {
    printf("Can't block SIGPIPE.\n");
    return eSystem;
  }

  sa.sa_flags = SA_SIGINFO;
  sa.sa_handler = &endit;
  if(sigaction(SIGINT, &sa, NULL))
  {
    printf("Can't handle SIGINT.\n");
    return eSystem;
  }

  printf("time,messages,data(K),errors,connections"
#ifdef USE_SSL
    ",SSL connections"
#endif
    "\n");

  log = new Logit("bhm.log", allLog, false, 0);
  Logit *debug = NULL;

  if(debugName)
    debug = new Logit(debugName, false, debugMultipleFiles, 0);

  int listen_fd = socket(PF_INET, SOCK_STREAM, 0);
  struct sockaddr_in in;
  in.sin_family = AF_INET;
  in.sin_port = htons(port);
  in.sin_addr.s_addr = INADDR_ANY;
  if(listen_fd == -1 || bind(listen_fd, (sockaddr *)&in, sizeof(in))
   || listen(listen_fd, 10))
  {
    fprintf(stderr, "Can't bind to port.\n");
    return 1;
  }
  struct pollfd pfd;
  pfd.fd = listen_fd;
  pfd.events = POLLIN;
  pthread_t *thread_info = (pthread_t *)calloc(sizeof(thread_info), maxThreads);
  thread_status = (int *)calloc(sizeof(int), maxThreads);
  time_t nextPrint = time(NULL)/60*60+60;
  while(1)
  {
    time_t t = time(NULL);
    if(t >= nextPrint)
    {
      res.print();
      nextPrint += 60;
    }
    int timeout = (nextPrint - t) * 1000;
    int rc = poll(&pfd, 1, timeout);
    if(rc < 0)
    {
      if(errno == EINTR)
        continue;
      fprintf(stderr, "Poll error on accept.\n");
      return 1;
    }
    int fd;
    struct sockaddr_in addr;
    socklen_t addrlen = sizeof(addr);
    if(rc == 1)
    {
      fd = accept(listen_fd, (sockaddr*)&addr, &addrlen);
      if(fd == -1)
      {
        res.error();
        continue;
      }
      res.connection();
      processes++;
      int i;
      for(i = 0; i < maxThreads && thread_status[i] == eUsed; i++)
        ;
      if(thread_status[i] == eFinished)
      {
        void *value_ptr;
        pthread_join(thread_info[i], &value_ptr);
        processes--;
      }

      thread_status[i] = eUsed;

      pthread_attr_t attr;
      if(pthread_attr_init(&attr)
       || pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE)
       || pthread_attr_setstacksize(&attr, 32*1024))
        fprintf(stderr, "Can't set thread attributes.\n");
      thread_data *td = (thread_data *)malloc(sizeof(thread_data));
      td->threadNum = i;
      td->fd = fd;
      memcpy(&td->addr, &addr, sizeof(addr));
      td->debug = debug ? new Logit(*debug, i) : NULL;
#ifdef USE_SSL
      td->ssl = use_ssl;
#endif
      int p = pthread_create(&thread_info[i], &attr, smtp_worker, PVOID(td));
      pthread_attr_destroy(&attr);
      if(p)
      {
        fprintf(stderr, "Can't create a thread.\n");
        return 1;
      }
    }
  }
  return 0;
}

